### PR TITLE
New check: Missing bridge structure (#341)

### PR DESCRIPTION
**Pull request by ladwlo**
### Description:

Port of Osmose check #7012. Flags long bridges without detailed info (bridge type/structure).
Issue https://github.com/osmlab/atlas-checks/issues/341

### Potential Impact:

None.

### Unit Test Approach:

Unit tests created for two positive scenarios (railway and major-highway bridges that meet all preconditions) and several negative scenarios (when some of the preconditions or check criteria are not met). Tests use local mini-atlases with artificial data.

### Test Results:

| ISO | Total Flags | Sampled  | Sampling % | TP | FP | False Positive Rate |
|-----|-------------|----------|------------|----|----| ------------------- |
| NZL | 24 | 24 | 100% | 24 | 0 | 0% |
| ARG | 62 | 20 | 33% | 20 | 0 | 0% |
| BOL | 3 | 3 | 100% | 3 | 0 | 0% |
| COL | 17 | 17 | 100% | 17 | 0 | 0% |
| DMA | 0 | 0 | 100% | 0 | 0 | 0% |
| CYM | 0 | 0 | 100% | 0 | 0 | 0% |
| TUR | 258 | 52 | 20% | 52 | 0 | 0% |

### CODE REVIEW

**Comment by ladwlo** [comment]
[BridgeDetailedInfoCheck - Analysis.pdf](https://github.com/osmlab/atlas-checks/files/5095599/BridgeDetailedInfoCheck.-.Analysis.pdf)
[BridgeDetailedInfoCheck - Requirements.pdf](https://github.com/osmlab/atlas-checks/files/5095602/BridgeDetailedInfoCheck.-.Requirements.pdf)

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/BridgeDetailedInfoCheck.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    @Override
+    protected List<String> getFallbackInstructions()
+    {
+        return FALLBACK_INSTRUCTIONS;
+    }
+
+    private boolean isGenericBridge(final AtlasObject object)
+    {
+        final Optional<String> bridgeTag = object.getTag(BRIDGE_TAG);
+        return bridgeTag.isPresent() && "yes".equals(bridgeTag.get());
```
Comment (Bentleysb) [inline src/main/java/org/openstreetmap/atlas/checks/validation/tag/BridgeDetailedInfoCheck.java:96]: This could be:
```suggestion
        return Validators.isOfType(object, BridgeTag.class, BridgeTag.YES);
```

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/BridgeDetailedInfoCheck.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    private boolean isMajorHighway(final AtlasObject object)
+    {
+        return MAJOR_HIGHWAYS.contains(((Edge) object).highwayTag());
+    }
+
+    private boolean isRailway(final AtlasObject object)
+    {
+        return object.getTag(RAILWAY_TAG).isPresent();
```
Comment (Bentleysb) [inline src/main/java/org/openstreetmap/atlas/checks/validation/tag/BridgeDetailedInfoCheck.java:106]: This could be:
```suggestion
        return Validators.hasValuesFor(object, RailwayTag.class);
```

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/BridgeDetailedInfoCheck.java (hunk 1 above)
Comment (ladwlo) [reply src/main/java/org/openstreetmap/atlas/checks/validation/tag/BridgeDetailedInfoCheck.java]: Done. Thanks for the hint!

File: config/configuration.json [hunk 3]
```diff
@@ -4,7 +4,7 @@
       "org.openstreetmap.atlas.checks.validation"
     ],
     "type": "org.openstreetmap.atlas.checks.base.BaseCheck",
-    "enabled.value.default": true
+    "enabled.value.default": false
```
Comment (sayas01) [inline config/configuration.json:7]: Please revert this change.

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/BridgeDetailedInfoCheck.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+     *
+     * @param object
+     *            the atlas object supplied by the Atlas-Checks framework for evaluation
+     * @return an optional {@link CheckFlag} object that
+     */
+    @Override
+    protected Optional<CheckFlag> flag(final AtlasObject object)
+    {
+        final Optional<String> bridgeStructureTag = object.getTag(BRIDGE_STRUCTURE_TAG);
+        if (((Edge) object).length().isGreaterThan(this.minimumLength)
```
Comment (sayas01) [inline src/main/java/org/openstreetmap/atlas/checks/validation/tag/BridgeDetailedInfoCheck.java:79]: Please use `this` keyword when accessing the class methods.

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/BridgeDetailedInfoCheck.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+public class BridgeDetailedInfoCheck extends BaseCheck<Long>
+{
+
+    private static final long serialVersionUID = -8915653487119336836L;
+
+    private static final EnumSet<HighwayTag> MAJOR_HIGHWAYS = EnumSet.of(HighwayTag.MOTORWAY,
+            HighwayTag.TRUNK, HighwayTag.PRIMARY, HighwayTag.SECONDARY);
+    private static final Double MINIMUM_LENGTH = 500.0;
+    public static final String BRIDGE_STRUCTURE_TAG = "bridge:structure";
+    private static final List<String> FALLBACK_INSTRUCTIONS = Arrays.asList(
```
Comment (sayas01) [inline src/main/java/org/openstreetmap/atlas/checks/validation/tag/BridgeDetailedInfoCheck.java:34]: nit: Since there is only a single instruction, you can use Collections.singletonList instead of Arrays.asList.

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/BridgeDetailedInfoCheck.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+     * @param object
+     *            the atlas object supplied by the Atlas-Checks framework for evaluation
+     * @return {@code true} if this object should be checked
+     */
+    @Override
+    public boolean validCheckForObject(final AtlasObject object)
+    {
+        // only master edges shall be flagged to avoid duplicate flags on the same OSM Way
+        return object instanceof Edge && ((Edge) object).isMasterEdge() && isGenericBridge(object)
+                && (isRailway(object) || isMajorHighway(object));
```
Comment (sayas01) [inline src/main/java/org/openstreetmap/atlas/checks/validation/tag/BridgeDetailedInfoCheck.java:65]: Looks like the RailwayTag class already has a public method isRailway which could be leveraged here like: RailwayTag.isRailway()

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/BridgeDetailedInfoCheck.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    private boolean isGenericBridge(final AtlasObject object)
+    {
+        return Validators.isOfType(object, BridgeTag.class, BridgeTag.YES);
+    }
+
+    private boolean isMajorHighway(final AtlasObject object)
+    {
+        return MAJOR_HIGHWAYS.contains(((Edge) object).highwayTag());
```
Comment (sayas01) [inline src/main/java/org/openstreetmap/atlas/checks/validation/tag/BridgeDetailedInfoCheck.java:101]: One other way to check this would be to use Validators class. Predicate can be declared as a constant at the top like:
`private static final Predicate<AtlasObject> MAJOR_HIGHWAYS = object -> Validators.isOfType(
            object, Highway.class, HighwayTag.MOTORWAY,
            HighwayTag.TRUNK, HighwayTag.PRIMARY, HighwayTag.SECONDARY)`

File: src/test/java/org/openstreetmap/atlas/checks/validation/tag/BridgeDetailedInfoCheckTestRule.java [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+    // short bridge
+    private static final String LOC_1 = "47.222,-122.444";
+    private static final String LOC_2 = "47.225,-122.441";
+    // long bridge
+    private static final String LOC_3 = "47.111,-122.666";
+    private static final String LOC_4 = "47.115,-122.661";
+
+    @TestAtlas(
+            // nodes
+            nodes = { @TestAtlas.Node(coordinates = @TestAtlas.Loc(value = LOC_3)),
```
Comment (sayas01) [inline src/test/java/org/openstreetmap/atlas/checks/validation/tag/BridgeDetailedInfoCheckTestRule.java:24]: nit: Could clean this class up by static importing TestAtlas.*.

File: src/test/java/org/openstreetmap/atlas/checks/validation/tag/BridgeDetailedInfoCheckTest.java [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+import org.openstreetmap.atlas.checks.configuration.ConfigurationResolver;
+import org.openstreetmap.atlas.checks.validation.verifier.ConsumerBasedExpectedCheckVerifier;
+import org.openstreetmap.atlas.utilities.configuration.Configuration;
+
+/**
+ * Tests for {@link BridgeDetailedInfoCheck}
+ *
+ * @author ladwlo
+ */
+public class BridgeDetailedInfoCheckTest
```
Comment (sayas01) [inline src/test/java/org/openstreetmap/atlas/checks/validation/tag/BridgeDetailedInfoCheckTest.java:15]: Love the extensive unit testing!! 💯

File: docs/checks/bridgeDetailedInfoCheck.md [hunk 10]
```diff
@@ -0,0 +1,25 @@
+# Bridge Detailed Info Check
```
Comment (sayas01) [inline docs/checks/bridgeDetailedInfoCheck.md:1]: Nice documentation!

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/BridgeDetailedInfoCheck.java [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+     * @param object
+     *            the atlas object supplied by the Atlas-Checks framework for evaluation
+     * @return an optional {@link CheckFlag} object that
+     */
+    @Override
+    protected Optional<CheckFlag> flag(final AtlasObject object)
+    {
+        final Optional<String> bridgeStructureTag = object.getTag(BRIDGE_STRUCTURE_TAG);
+        if (((Edge) object).length().isGreaterThan(this.minimumLength)
+                && !bridgeStructureTag.isPresent())
```
Comment (sayas01) [inline src/main/java/org/openstreetmap/atlas/checks/validation/tag/BridgeDetailedInfoCheck.java:80]: Could use bridgeStructureTag.isEmpty() instead.

File: config/configuration.json (hunk 3 above)
Comment (ladwlo) [reply config/configuration.json:7]: Done.

Comment (ladwlo) [reply config/configuration.json]: Thanks for all the detailed remarks!
